### DISCUSSION TITLE
Question: Access/Match Last Consumed Character(s) of Input

### DISCUSSION THREAD

**Discussion by sanssecours** [taocpp/PEGTL#144] [Q&A] [answered]
# Issue Description

I wrote a [very basic parser](https://github.com/sanssecours/YAy-PEG) for a small subset of [YAML](https://yaml.org) using PEGTL. While I was able to construct rules for most of the [YAML specification](https://yaml.org/spec/1.2/spec.html) without too much problems, there is a particular grammar rule that my code currently does not handle. The spec defines the rule [`ns-plain-char(c)`](https://yaml.org/spec/1.2/spec.html#ns-plain-char(c)) as:

```c
ns-plain-char(c) ::= (ns-plain-safe(c) - ":" - "#")               | 
                     (/* An ns-char preceding */ "#")             | 
                     (":" /* Followed by an ns-plain-safe(c) */ )
```

. The problem here is the comment “An `ns-char` preceding”. This part of the rule specifies that the **last consumed character** should match the rule [`ns-char`](https://yaml.org/spec/1.2/spec.html#ns-char). 

One option to work around this issue would be to create a custom rule that checks the last character in the input. However, as far as I can tell, the input class only provides methods to check the text the parser did not consume yet (`peek_char`, `peek_byte`). Do you have an idea how I can solve this problem without too much hassle?

**Comment by ColinH** [comment, marked as answer]
Scanning backwards is not properly supported because the usual (?) way with PEGs is to make use of the infinite look-ahead and disambiguate in advance. Or you rework the grammar in other ways, by deferring decisions to as late as possible.

Now you would presumable prefer to use the YAML grammar as is, without major refactoring.

The `memory_input` now also remembers where it started parsing, so you could actually check whether there are bytes between `begin()` and `current()`, and implement a kind of backward-scanning UTF-8 decoding function.

So it's a hassle, but it's possible, and I couldn't say whether there is a more elegant solution without thinking about it some more.

**Comment by sanssecours** [comment]
Thank you for the detailed answer. I [used the second approach](https://github.com/sanssecours/YAy-PEG/commit/21a5d58ef4c041ea373c135059c6918c025f102e) you described in my project. As far as I can tell it seems to work.

**Comment by ColinH** [comment]
How are your bridging the gap between the input's `current()` method that usually returns a `const char*` and knows nothing about UTF-8, and the 21/32-bit code units you are comparing with?

**Comment by sanssecours** [comment]
I guess I do not do that 😄. I just assumed `current()` returns a pointer to the last character in UTF-32 form. At least that would be what it would return in my perfect version of this world.

Is there an easy method to handle the encoding?

**Comment by sanssecours** [comment]
I changed the rule `ns_char_preceding` in my [latest commit](https://github.com/sanssecours/YAy-PEG/commit/477189f858c4790bd4ddf32f7eb53d2124719abe). Now the rule should work correctly, if the input is encoded in UTF-8.

**Comment by ColinH** [comment]
There's no easy way (that we supply), glad that you managed to make it work (even if the hard way).
